### PR TITLE
Tdd

### PR DESCRIPTION
**Pull request by sehhajdkg**

### CODE REVIEW

**Comment by sehhajdkg** [comment]
1. Completed Intro tutorial (https://github.com/sehhajdkg/step/blob/master/walkthroughs/week-5-tdd/intro/tdd-intro-walkthrough.md)
2. query function in FindMeetinQuery.java - see if request for new meeting can validated. (https://github.com/sehhajdkg/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,98 @@
...
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    //If there are no events or no one is attending the event, 
+    // all requests are possible if they are within the range of the whole day.
+    if(request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+      if(events.isEmpty() || request.getAttendees().isEmpty()) {
```
Comment (bldouglas) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: minor nit here, but it feels cleaner to minimize nesting and separate out the different conditions (similar thing applies below as well). For example:
`if (duration > whole day) {
  return empty list
}
if (no events / attendees) {
  return whole day singleton
}
`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,98 @@
...
+    List<TimeRange> potentialTime = new ArrayList<>();
+
+    if(bookedEvents.isEmpty()){
+      // attendees are not booked for any events, are available for the whole day.
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } else {
+
+      Event prevEvent = bookedEvents.get(0); 
+      Event currEvent = bookedEvents.get(0);
+      for(int i = 0; i < bookedEvents.size(); i++){
```
Comment (bldouglas) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I'd consider pulling out a separate method for this inner loop (for example, passing through the current 'tail' potential time period (initialized to the whole day) along with the next booked event, and returning the 0 to 2 replacement time periods not overlapping with that event.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,98 @@
...
+
+        if(potentialTime.size() == 0) { 
+          // See if there is a slot before the first booked event and if this slot.
+          // is large enough for the request. 
+          if(currEvent.getWhen().start() != TimeRange.START_OF_DAY) {
+            // If request duration is within slot between start of day and first event.
+            if((TimeRange.START_OF_DAY+request.getDuration()) <= currEvent.getWhen().start()){
+              potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+            } 
+            potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));
```
Comment (bldouglas) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: does this handle the situation where currEvent ends less than request.getDuration() from the end of the day?

Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Added in this case

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,185 @@
...
+
+    // Find all events that attendees in the meeting request are already booked for.
+
+    // Mandatory attendees.
+    List<Event> mandatoryBookedEvents = findBookedEvents(events, request.getAttendees());
+   
+    // Optional attendes.
+    List<Event> optionalBookedEvents = findBookedEvents(events, request.getOptionalAttendees());
+
+    // All attendees - Mandatory + Optional.
```
Comment (waxmittmann) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Is this an old comment? It doesn't seem to fit the code that follows it.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,185 @@
...
+    List<TimeRange> potentialTime = new ArrayList<>();
+    
+    Event prevEvent = bookedEvents.get(0); 
+    Event currEvent = bookedEvents.get(0);
+    for(int i = 0; i < bookedEvents.size(); i++){ 
+      prevEvent = currEvent; 
+      currEvent = bookedEvents.get(i); 
+
+      // If only one event - consider only timerange before and after the event.
+      if(bookedEvents.size() == 1) {
```
Comment (waxmittmann) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: There's no reason for this block to be inside the loop and having it here makes the code much harder to read since the reader needs to understand that this code will only potentially trigger once, during the first iteration. Put it before the loop like so:
```
if (bookedEvents.size() == 1) { 
        potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(), false));
        potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
        return potentialTime;
}
```

This is called an **early exit**; based on a condition (bookedEvents.size() == 1) we know that we don't have to perform the main body of logic.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,185 @@
...
+        potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(), false));
+        potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+        break;
+      }
+
+      // No times found yet.
+      if(potentialTime.size() == 0) {
```
Comment (waxmittmann) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Since the following two 'if' blocks don't have an 'else' and don't do logic outside the if block, it's much more readable to write this as a single block: 
`if (potentialTime.size() == 0 && currEvent.getWhen().start != TimeRange.START_OF_DAY && (TimeRange.START_OF_DAY+request.getDuration()) < currEvent.getWhen().start())`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,185 @@
...
+        // See if there is a slot before the first booked event and if this slot.
+        // is large enough for the request. 
+        if(currEvent.getWhen().start() != TimeRange.START_OF_DAY) {
+          // If request duration is within slot between start of day and first event (curr).
+          if((TimeRange.START_OF_DAY+request.getDuration()) < currEvent.getWhen().start()){
+            potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+          }
+        }
+        // If meeting request can fit in between end of event and end of day.
+        if((currEvent.getWhen().end()+request.getDuration()) <= TimeRange.END_OF_DAY){
```
Comment (waxmittmann) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: If I read the code correctly this works because the 'else' block re-checks the last potential time to see if it conflicts. Having to keep that much context in mind makes it hard to follow the logic. 

Rather than having a special case for `potentialTime.size() == 0` it would probably make more sense to just process the first booked event before the loop and then start the loop at the second event.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,185 @@
...
+          }
+          potentialTime.add(TimeRange.fromStartEnd(laterEnd, TimeRange.END_OF_DAY,true));
+        } else {
+          // check GAP (...) in-between curr and prev.
+          // |--prev--|...|--curr--|
+          long gapDuration = currEvent.getWhen().start() - prevEvent.getWhen().end();
+          if(gapDuration >= request.getDuration()){
+            potentialTime.add(TimeRange.fromStartEnd(prevEvent.getWhen().end(), currEvent.getWhen().start(), false));
+          }
+          // Check there there is a gap between the end of the event and end of day large enough for request.
```
Comment (waxmittmann) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: If you compare an event against the previous event (which is always possible if you process the first event outside the loop, then there will always be a previous event) then assuming that the gap between them is big enough you can just add it to potentialTime without having to re-check it again later. This will allow you to remove a lot of logic, significantly reducing complexity.

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 9]
```diff
@@ -22,6 +22,13 @@
    * Returns a greeting for the given name.
    */
   public String greet(String name) {
-    return "Hello " + name;
+    return "Hello " + removeNonAlphabetChar(name.trim());
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java]: I believe trim removes whitespace which is a non-alphabet character so this call is redundant.

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 10]
```diff
@@ -22,6 +22,13 @@
...
+  }
+
+  /**
+   * Returns name after removing all characters that are not alphabet letters.
+   */
+  public String removeNonAlphabetChar(String name){
+    return name.replaceAll("[^a-zA-Z]", "");
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java]: having said the above this code doesn't allow you to have two names, eg "Lauren Spooner" would become "LaurenSpooner" so maybe keeping the trim and allowing spaces in here is better?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,185 @@
...
+   * Finds all events that mandatory and optional attendees are booked for - removes duplicates.
+   * @param mandatory - events that mandatory attendees in the request have to attend.
+   * @param optional - events that optional attendees in the request have to attend.
+   * @return list of events.
+   */
+  public List<Event> getAllUniqueEvents(List<Event> mandatory, List<Event> optional) {
+
+    List<Event> allEvents = new ArrayList<>(mandatory);
+
+    for(Event opt : optional){
```
Comment (waxmittmann) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit:  What's the difference between an 'optional' and an 'opt'? The naming makes it hard to understand what's going on here.

The reason you chose 'opt' is because you've named the inputs mandatory and optional, which hides the fact that they are lists. Best to use plural in this case, like you did with 'allEvents'. If you rename the input arguments to 'mandatoryEvents' and 'optionalEvents' then 'opt' can become 'optionalEvent'.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,185 @@
...
+      }
+    }
+    return allEvents;
+  } 
+  /**
+   * Basic Bubble sorting of event list so they are in ascending order by start time.
+   * @param events - list of events to be compared by start time.
+   * @return sorted list of events.
+   */
+  public List<Event> sortEventsByStartTime(List<Event> events) {
```
Comment (waxmittmann) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:181]: While it's nice to see you got some practice writing a bubble sort, `Collections.sort` should get the job done without having to implement your own sorting algorithm. It's usually a good idea to use a library sorting function because it's easy to introduce subtle errors when rolling your own.

**Review by waxmittmann — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -121,6 +123,106 @@ public void everyAttendeeIsConsidered() {
...
+    // Now only the early and late parts of the day should be returned. 
+    // (Times when All (Optional/Mandotory) Attendees available)
+
+    // Have each person have different events. We should see three options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|     |--B--|
+    //                        |-C-|
+    // Day     : |-----------------------------|
+    // Options : |--1--|     |--2--|     |--3--|
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I think this option is gone now

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -14,10 +14,185 @@
...
+    // All attendees - Mandatory + Optional.
+    List<Event> allBookedEvents = getAllUniqueEvents(mandatoryBookedEvents, optionalBookedEvents);
+    allBookedEvents = sortEventsByStartTime(allBookedEvents);
+
+    // No possible meeting if request is longer than a day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+    // No events/no attendess means meeting can be anytime.
+    if(events.isEmpty() || (mandatoryBookedEvents.isEmpty() && optionalBookedEvents.isEmpty())) {
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Is it possible for the mandatory and optional events to be non-empty if the events list is empty? or visa versa?
(To ask this question another way are you just doing the same check twice?)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -14,10 +14,185 @@
...
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.List;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    //If there are no events or no one is attending the event,
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This seems like it should be a function comment (ie above the function like the javadoc comments) otherwise I would say the comments should be close to the code that's doing the things the comment is talking about, though you seem to have commented the rest of the function well.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,185 @@
...
+    //If there are no events or no one is attending the event, 
+    // all requests are possible if they are within the range of the whole day.
+
+    // The basic functionality of optional attendees is that if one or more time slots exists 
+    // so that both mandatory and optional attendees can attend, return those time slots. 
+    // Otherwise, return the time slots that fit just the mandatory attendees.
+
+    // Find all events that attendees in the meeting request are already booked for.
+
+    // Mandatory attendees.
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: you probably don't need comments explaining these variables, the variable name itself does that

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,185 @@
...
+  } 
+  /**
+   * Finds all potential availabilities in between bookedEvents.
+   * @param bookedEvents - timerange already booked out/unavailable.
+   * @param request - meetin request to add in a potential new timerange.
+   * @return list of timeranges for availability.
+   */
+  public Collection<TimeRange> findPotentialTimes(List<Event> bookedEvents, MeetingRequest request){
+    // List of potential free timeranges for the meeting request.
+    List<TimeRange> potentialTime = new ArrayList<>();
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Your function returns a collection so this should probably be a collection type, unless there's some particular list property you need in this function.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -14,10 +14,185 @@
...
+    
+    Event prevEvent = bookedEvents.get(0);
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:122]: What is bookedEvents is empty?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -14,10 +14,185 @@
...
+    
+    Event prevEvent = bookedEvents.get(0); 
+    Event currEvent = bookedEvents.get(0);
+    for(int i = 0; i < bookedEvents.size(); i++){ 
+      prevEvent = currEvent; 
+      currEvent = bookedEvents.get(i); 
+
+      // If only one event - consider only timerange before and after the event.
+      if(bookedEvents.size() == 1) {
+        potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(), false));
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: what if there's not enough time for the meeting between the start or the end of the day?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -14,10 +14,185 @@
...
+        break;
+      }
+
+      // No times found yet.
+      if(potentialTime.size() == 0) { 
+        // See if there is a slot before the first booked event and if this slot.
+        // is large enough for the request. 
+        if(currEvent.getWhen().start() != TimeRange.START_OF_DAY) {
+          // If request duration is within slot between start of day and first event (curr).
+          if((TimeRange.START_OF_DAY+request.getDuration()) < currEvent.getWhen().start()){
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: this check will fail if the event starts at the beginning of the day so I don't think you need the if statement above.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -14,10 +14,185 @@
...
+            potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+          }
+        }
+        // If meeting request can fit in between end of event and end of day.
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: rather than adding and removing this end of day timerange for every loop it might be easier to just do a check after the for loop for everything from the last meeting to the end of the day.
That way you could probably remove a bunch of the special cases in here too, you just need to check between the current and previous event every time (with a special case for i = 0) and then have a final if statement after hte for loop.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,185 @@
...
+
+    for(Event opt : optional){
+      if(!allEvents.contains(opt)){
+        allEvents.add(opt);
+      }
+    }
+    return allEvents;
+  } 
+  /**
+   * Basic Bubble sorting of event list so they are in ascending order by start time.
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: bubble sort is not a great sorting algorithm, and we very rarely write our own sorts anyway when most languages implement them for us.
you can simply use events.sort(<lambda comparator>) where lambda comparator is a lambda function that compares the properties you care about in the list objects, so in this case the event start time.

**Review by lorikeet-s — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 18 above)
Comment (waxmittmann) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:122]: That's a good point. As the function is currently used there's a check to make sure it's not before the function is called, but that may not always be the case. Someone might come along and reuse the function, or refactor the code that calls it in such a way that the list might be empty.

A function should always handle any 'legal' input correctly. You can declare certain inputs 'illegal' by stating preconditions in the function comment, e.g. "Preconditions: 'bookedEvents' must be non-empty", but you should always check all those preconditions in the functions and throw an informative error if they aren't met.

For collections the caller will generally expect a function to be able to handle an empty list (a common case) so you should usually handle this without raising an error, i.e. by adding an early exit that returns a sensible value immediately if the input list is empty.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 22 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 20 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 19 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 18 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:122]: Done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 15 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 14 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 13 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Removed it

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java (hunk 10 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:181]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 21 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -99,100 +94,93 @@
...
           long gapDuration = currEvent.getWhen().start() - prevEvent.getWhen().end();
           if(gapDuration >= request.getDuration()){
             potentialTime.add(TimeRange.fromStartEnd(prevEvent.getWhen().end(), currEvent.getWhen().start(), false));
           }
-          // Check there there is a gap between the end of the event and end of day large enough for request.
-          if(currEvent.getWhen().end() != TimeRange.END_OF_DAY){
-            if((currEvent.getWhen().end()+request.getDuration()) <= TimeRange.END_OF_DAY){
-              potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));  
-            } 
+          // Last event in list then check there there is a gap between
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: typo

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -21,42 +21,37 @@
 import java.util.Collections;
 
 public final class FindMeetingQuery {
+  /**
+   * The basic functionality of query is that if one or more time slots exists 
+   * so that both mandatory and optional attendees can attend, return those time slots. 
+   * Otherwise, return the time slots that fit just the mandatory attendees.
+   * @param events
+   * @param request
+   * @return
```
Comment (lorikeet-s) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: returns what?

**Review by lorikeet-s — CHANGES_REQUESTED** [review]
Looks good, that loop is much cleaner now, couple of minor issues with the comments but otherwise it's good to go.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 23 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 24 above)
Comment (sehhajdkg) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Added comment for return

**Review by lorikeet-s — APPROVED** [review]
